### PR TITLE
Bind extensions in initializers of inferred locals

### PR DESCRIPTION
**Pull request by jcouv**
Relates to test plan https://github.com/dotnet/roslyn/issues/66722

### CODE REVIEW

**Comment by AlekseyTs** [comment]
>     public void ExtensionMemberLookup_TypeOnlyContext_ExtensionTypeOnUnderlying_NonGeneric()

I am not sure what is going on here. Why the existing tests are getting modified this way? If we are deleting some tests that are no longer relevant and adding new tests, then why don't we add new tests at the end of the file to avoid any confusion for the reader?

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:13512 in e9c7eb9. [](commit_id = e9c7eb979871a09724d8333b41fd557b7633915e, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     public void ExtensionMemberLookup_AsFunctionType_StaticMethod_WithTypeArgument()

It is not obvious why this chunk of tests is getting removed

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:13669 in e9c7eb9. [](commit_id = e9c7eb979871a09724d8333b41fd557b7633915e, deletion_comment = True)

**Comment by AlekseyTs** [comment]
>     public void ExtensionInvocation_Instance_ExtensionOnUnderlying()

What "ExtensionOnUnderlying" in the name of the test is supposed to convey? We are accessing ```M``` on ```E```, there is no other ```M``` in the picture. I assume the result doesn't matter whether the access happens in ```C``` or not. So what unique property of the scenario we are trying to convey?

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:16472 in e9c7eb9. [](commit_id = e9c7eb979871a09724d8333b41fd557b7633915e, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     public void InferredVariable_Instance_StaticMethod()

I am unable to make sense out of the diff in a reasonable amount of time. From the diff it appears that existing tests are renamed, and scenarios in them are changed. The motivation to do this is not obvious.

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:17810 in e9c7eb9. [](commit_id = e9c7eb979871a09724d8333b41fd557b7633915e, deletion_comment = False)

**Comment by AlekseyTs** [comment]
I am really confused about what is going on with tests in this PR. It appears as though tests for scenarios unrelated to the stated goal of the PR are getting added/removed/changed, or, perhaps even moved. Is there a way to separate things so that it would be clear for a reviewer what the intent is and why the intent makes sense? Perhaps modifications, additions for tests targeting scenarios not primarily related to the stated goal of the PR could be separated into a dedicated PR? Perhaps the changes could be organized in such a way that the diff is easier to follow, like it wouldn't appear as though we rename and at the same time drastically change tests scenarios in existing tests? For now, I stopped reviewing this PR.

**Comment by jcouv** [comment]
>     public void ExtensionMemberLookup_TypeOnlyContext_ExtensionTypeOnUnderlying_NonGeneric()

I tried doing a few things (removing some duplicate tests, clarifying/aligning the name of some existing tests, adding some tests). That did make the diff pretty bad.
It should be better now

---
In reply to: [1985051781](https://github.com/dotnet/roslyn/pull/72398#issuecomment-1985051781) [](http://example.com/codeflow?ancestors=1985051781)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:13512 in e9c7eb9. [](commit_id = e9c7eb979871a09724d8333b41fd557b7633915e, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         Assert.Null(model.GetSymbolInfo(memberAccess).Symbol);

Consider checking ```CandidateSymbols``` and ```CandidateReason```

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:13594 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     [ConditionalFact(typeof(CoreClrOnly))]

Why is this necessary?

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:13672 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     [ConditionalFact(typeof(CoreClrOnly))]

Why is this necessary?

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:13512 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     [ConditionalFact(typeof(CoreClrOnly))]

Why is this necessary?

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:10666 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     [ConditionalFact(typeof(CoreClrOnly))]

Why is this necessary?

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:10698 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     [ConditionalFact(typeof(CoreClrOnly))]

Why is this necessary?

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:10698 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     [ConditionalFact(typeof(CoreClrOnly))]

> Why is this necessary?

Is usage of ```TargetFramework.Net70``` required? If so, for other features we, usually, simply use conditional logic in ```CompileAndVerify``` to suppress execution for desktop. 

---
In reply to: [1985951519](https://github.com/dotnet/roslyn/pull/72398#issuecomment-1985951519) [](http://example.com/codeflow?ancestors=1985951519)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:13672 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         Assert.Null(model.GetSymbolInfo(memberAccess).Symbol);

CandidateSymbols? CandidateReason?

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:18056 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -408,6 +408,23 @@ internal BoundExpression BindToNaturalType(BoundExpression expression, BindingDi
...
+            CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+            MethodGroupResolution resolution = ResolveMethodGroup(methodGroup, analyzedArguments: null, isMethodGroupConversion: false, ref useSiteInfo);
+            diagnostics.Add(methodGroup.Syntax, useSiteInfo);
+            if (resolution.IsExtensionMember(out Symbol? extensionMember))
+            {
+                return GetExtensionMemberAccess(methodGroup.Syntax, methodGroup.ReceiverOpt, extensionMember, diagnostics);
+            }
+
+            // We have a method group so bind to an inferred delegate type
+            return BindToInferredDelegateType(methodGroup, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:424]: >return BindToInferredDelegateType(methodGroup, diagnostics); [](http://example.com/codeflow?start=12&length=60)

Given that ```IsExtensionMember``` really means ```IsNonMethodExtensionMember```, is ```MethodGroup``` going to be mutated by ```ResolveMethodGroup``` when we are dealing with extension methods? It looks somewhat surprising that ```resolution``` isn't involved on this code path.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -10204,17 +10212,51 @@ static bool isCandidateUnique(ref MethodSymbol? method, MethodSymbol candidate)
...
                     foreach (var extensionMethod in methodGroup.Methods)
                     {
                         var substituted = typeArguments.IsDefaultOrEmpty ? extensionMethod : extensionMethod.Construct(typeArguments);
 
-                        var reduced = substituted.ReduceExtensionMethod(receiver.Type, Compilation, out bool wasFullyInferred);
+                        var type = receiver.Type;
+                        Debug.Assert(type is not null);
+                        if (type.ExtendedTypeNoUseSiteDiagnostics is { } extendedType)
+                        {
+                            type = extendedType;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:10256]: >type = extendedType; [](http://example.com/codeflow?start=28&length=20)

Why is this the right thing to do? What if the extension method extends the extension type? What if it extends a type parameter?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 3]
```diff
@@ -10289,6 +10329,51 @@ bool satisfiesConstraintChecks(MethodSymbol method)
...
+
+                    Symbol symbol = GetSymbolOrMethodOrPropertyGroup(lookupResult, node, rightName, arity, members,
+                        BindingDiagnosticBag.Discarded, wasError: out _, qualifierOpt: null);
+
+                    // Ignore non-method symbols
+                    if (symbol is null)
+                    {
+                        Debug.Assert(members.Count > 0);
+                        Debug.Assert(members[0].Kind is SymbolKind.Method);
+                        methodGroup.PopulateWithMethods(left, members, typeArgumentsWithAnnotations, isExtensionMethodGroup: false, resultKind: lookupResult.Kind);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:10369]: >PopulateWithMethods [](http://example.com/codeflow?start=36&length=19)

Is it safe to call this method multiple times? Is it going to discard results from the previous call?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:10369]: We call `methodGroup.Clear();` before calling `populateExtensionTypeMethodsFromSingleBinder`, so we're always populating into a clear method group.
The logic in `PopulateWithMethods` has assertions to that effect as well (only can call populate on a clear method group, see `VerifyClear()`).

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:10256]: > What if the extension method extends the extension type?

That's not allowed:
> We modify the extension methods rules as follows:
> [...] The first parameter of an extension method may have no modifiers other than this, and the parameter type may not be a pointer or an extension type.

> What if it [the extension method] extends a type parameter?

I think that's only possible in error scenarios, as extension methods must be defined in a top-level non-generic class.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:10256]: >> What if it [the extension method] extends a type parameter?

> I think that's only possible in error scenarios, as extension methods must be defined in a top-level non-generic class.


For example, this isn't an error scenario:
```
static class Ext
{
    public static T M<T>(this T a) => a;
}
```

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:10256]: > We modify the extension methods rules as follows

Consider capturing the fact in comments, to make motivation for the logic clear.

**Comment by jcouv** [comment]
>     [ConditionalFact(typeof(CoreClrOnly))]

Without `TargetFramework.Net70` we get a compilation error ("missing CompilerFeatureRequiredAttribute")

---
In reply to: [1985972932](https://github.com/dotnet/roslyn/pull/72398#issuecomment-1985972932) [](http://example.com/codeflow?ancestors=1985972932,1985951519)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:13672 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     [ConditionalFact(typeof(CoreClrOnly))]

> missing CompilerFeatureRequiredAttribute"

We are not synthesizing its definition? This doesn't feel right. If that is the only thing that prevents us from running on desktop, we should remedy this.


---
In reply to: [1986331743](https://github.com/dotnet/roslyn/pull/72398#issuecomment-1986331743) [](http://example.com/codeflow?ancestors=1986331743,1985951519)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:13672 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

**Comment by jcouv** [comment]
>     public void ExtensionInvocation_Instance_ExtensionOnUnderlying()

The reason why a lookup of "M" on a type `E1` resolves is because we have an extension `E2.M` which applies to `C` (the underlying type).

---
In reply to: [1985057710](https://github.com/dotnet/roslyn/pull/72398#issuecomment-1985057710) [](http://example.com/codeflow?ancestors=1985057710)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:16472 in e9c7eb9. [](commit_id = e9c7eb979871a09724d8333b41fd557b7633915e, deletion_comment = False)

**Comment by jcouv** [comment]
>     [ConditionalFact(typeof(CoreClrOnly))]

I have a task list item to ensure we can run static scenarios on desktop. I'll leave a PROTOTYPE comment here as well.

---
In reply to: [1986352334](https://github.com/dotnet/roslyn/pull/72398#issuecomment-1986352334) [](http://example.com/codeflow?ancestors=1986352334,1985951519)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:13672 in 24afadf. [](commit_id = 24afadfb18ecc19a3492a7dce13060edaca5e525, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:424]: > is MethodGroup going to be mutated by ResolveMethodGroup when we are dealing with extension methods?

I don't see how `ResolveMethodGroup` could mutate the `BoundMethodGroup` parameter. The members of `BoundMethodGroup` are readonly properties and the collections are readonly (`ImmutableArray`).

> It looks somewhat surprising that resolution isn't involved on this code path.

Yes, at the moment, we have separate logic for resolving method groups and for determining natural function type. 

FWIW, I have a related open design issue on this, where an instance method with bad arity could hide a good extension field. It's possible we could arrive at a way of unifying those two processes.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 6)

**Comment by jcouv** [comment]
@jjonescz @dotnet/roslyn-compiler for second review. Thanks

**Review by jjonescz — APPROVED** [review]

**Pull request merged**